Major docs rebuild

zsup opened the pull request:
This pull request includes a large number of changes. In particular, it includes:
- A dramatic reorganization of the content of the docs in order to support more "linear" guides (see the "How to build a product with Particle" section)
- A rebuild of the docs using [Metalsmith](http://www.metalsmith.io) instead of [Assemble](http://www.assemble.io)
- A better test suite that ensures there are no broken links, both internal and external (currently these tests are failing all over the place because ALL the links are broken)

As of the creation of the pull request, there is still a lot of work to do for this to be ready. I expect that this will take another few days of concerted development effort, both on the content creation side and the development side. Here's a list of things still to be done:

Content:
- [ ] Finish 'How to build a product with Particle'
- [ ] Split some of the guide sections into separate 'guides' and 'reference' (a lot of them combine both)
- [ ] Fix all the links that have been broken

Development:
- [x] Rebuild the navigation (both the top and the sidebar)
- [x] Create navigation across guide sections (i.e. click 'next' and 'previous', like Gitbook)
- [x] Properly utilize Metalsmith's 'collections' for organizing sections of content
- [ ] Performance improvements on the big pages (make the 'firmware' reference load faster and not break mobile devices)
- [x] Redo responsive styling
- [ ] New search (powered by Lunr and metalsmith-lunr)
- [x] Rebuild the .zip download (removed this feature for now) 
- [ ] Set up a whole boatload of redirects so old links don't break
- [ ] Set up a nice 404 page
- [ ] Move all of the content from the 'support' site (Zendesk) into its own collection
- [x] Fix Jeff's bug in the comments below
- [ ] Ensure the tests pass
- [ ] Create a hamburger menu for mobile where the sidebar content goes

There's probably more; feel free to add whatever I may have missed.

jme783 commented:
Was running into an issue when trying to get this to build locally on my machine. When running `npm start`, I would see:

```
Error: Failed to read the file at: /Users/Jeff/Documents/particle-apps/docs/src/assets/images/mk-diode.bmp

EMFILE, open '/Users/Jeff/Documents/particle-apps/docs/src/assets/images/mk-diode.bmp’
```

To fix this issue, I had to add `ulimit -n 2560` to my `bash_profile`, as outlined here: https://github.com/substack/node-browserify/issues/431

cmsunu28 committed:
latest content updates

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Highlight the last subheader of the previous section when you scroll back up

jme783 committed:
Remove h4s from the TOC

jme783 commented:
@zsup fixed issue with too many files open concurrently

jme783 committed:
Made good progress on the dashboard guide

zsup committed:
Fixed redirects at the root of the site and in redirect.js

zsup committed:
Renamed all the 'how to build a product' pages

zsup committed:
Content weaks to intro and conclusion for 'how to build a product'

cmsunu28 committed:
10 Content Changes for your 4th of July Weekend

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Wahoo CSS! Make the guide more responsive

cmsunu28 committed:
Six Amazing Content Updates to Fix Your CSS Errors

jme783 committed:
Added redirects for the main three sections of the guide

jme783 committed:
Big commit -- separated out the templates

Commit:
Outlines for PCB design and manufacturing

Added some outlines, made bold the sections beyond what Zach laid out
in his first pass.

cmsunu28 committed:
New CSS Pop-Ups: You'll Either Love Them, or Love to Hate Them

cmsunu28 committed:
This Image Popup Fix Will Make You Hate the Popups Less

idokleinman committed:
more progress on how to build your mobile app - specifically iOS

jme783 committed:
Evaluate special characters correctly when generating the TOC

jme783 committed:
Fixed issue with code not indenting correctly

jme783 committed:
Refactored scrollspy to only create h3 waypoints when the corresponding h2 has been scrolled to

jme783 committed:
created smooth animation for scrolling to content

jme783 committed:
Improve in page toc in the reference section

jme783 committed:
Continue improving menu experience

jme783 committed:
Cleaned up Scrollspy code and fixed a repainting issue on chrome

jme783 committed:
Added instructions on nav structure to the READme

jme783 committed:
Fixed issue with arrow covering the scrollbar on the guide

jme783 committed:
Removed console log statements

jme783 committed:
Clean up some cruft and go back to using metalsmith-watch. Now live reload works with css changes, yay

jme783 committed:
Added a template for datasheets

cmsunu28 committed:
Non-technical team members love these new media helpers! Find out why here.

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

cmsunu28 committed:
Vine Helpers: Now With Working Video

jme783 committed:
Fleshing out dashboard content on How to Build a Product, and fixed 404 with old reference to stylesheets

jme783 committed:
Add lunr package and added lunr: true to metadata of every md file

jme783 committed:
Continue fleshing out dashboard docs page in the Guide

jonlogan committed:
Removed the Callback section from Events because it contained incorrect date and feature information

jme783 committed:
Generate A Luna manifest file on a section-by-section basis

jme783 committed:
Started styling search=

jme783 committed:
Add search to the docs!

jme783 committed:
Use handlebars to do lunr search!

jme783 committed:
Use metalsmith-register-partials and clean up templates. Yes.

Commit:
Added Support page
Added support page images to src/assets/images/
Added new template page templates/support.hbs
Beginning to creat new support-popup.less file

Commit:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

cmsunu28 committed:
New and Exciting Content Changes, Only .99

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Added order to front matter

jme783 committed:
Remove numbers from the titles of guide pages, and make use of ordered lists

jme783 committed:
styled the no results for the search

jme783 committed:
Allow for sorting of dynamic collections for use on the guide page

jme783 committed:
Fixed {{assets}} link

jme783 committed:
Fixed bug with tracking on click

cmsunu28 committed:
Content Changes to Make Your iOS SDK Organized for the Summer

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Clean up the Cloud API page

jme783 committed:
Clean up getting started markdown

cmsunu28 committed:
Changes to streamline the lovely datasheet page

cmsunu28 committed:
Removed numbering system from Photon and P1 datasheets

chobberoni committed:
Added backertkit support page
Modified the support.hbs file to support different less files
Updated the metalsmith.js file to add support to its server exports
New images added for shipping-account-info.md
New support-popup.js and support-popupp.less files that will be later
merged into popup.js and popup.less

chobberoni committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

cmsunu28 committed:
Making footnote.less and popup.less work better on mobile

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Drastically change how IDs and scrollspy work to accommodate duplicates

jme783 committed:
Scroll in page when search result is hit

brycekahle committed:
Add section title to head partial

brycekahle committed:
Add initial API doc generation

brycekahle committed:
remove api-node submodule

brycekahle committed:
readd submodule on https and handle PRs that can't check it out

brycekahle committed:
cd out of dir after success submodule update

jme783 committed:
Fixed merge conflicts with metalsmith.js

chobberoni committed:
Revert "Add section title to head partial"

This reverts commit abbcd15b4da71c243b6e6023e6f64eb96a6ae0b6.

chobberoni committed:
Revert 0b6e92d..74c2557

This rolls back to commit 0b6e92dd4d27ade9631b007889f43b705ab6a4f2.

brycekahle committed:
Fix apidocs plugin when submodule hasn't been initialized

jme783 committed:
Keep pushing the How to Build a Product -- Dashboard page forward

jme783 committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith

brycekahle committed:
Move api doc gen to a partial so we can indent. Support for additional operations per endpoint.

idokleinman committed:
how to create mobile app guide major progress

brycekahle committed:
Fix TOC for generated content. Assign order to routes.

towynlin committed:
Web app guide in progress, solid first pass

jme783 committed:
Finished Dashboard docs content for How to Build a Product

jme783 committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith

jme783 committed:
Fix scrollspy on reference, support, and datasheets

jme783 committed:
Fix internal links

jme783 committed:
expand the table of contents when the user reaches the first header

jme783 committed:
Cleaned up support template

chobberoni committed:
New backerkit information

new backerkit .md file
new backer kit login images
updated support template
changed the order of the support page

jme783 committed:
Improved kits page

jme783 committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith

jme783 committed:
Removed stashed changes mumbo jumbo

jme783 committed:
Fixed logic with +/-

jme783 committed:
Increased timeout of 2 tests to get them to pass

cmsunu28 committed:
added redirects

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
moved scripts from the header to below the body for performance gains

chobberoni committed:
Moving on to last stage: Troubleshooting docs

Finished kickstarter and other layout issues.
fixed al code breaks that i was having.
now on to Troubleshooting stages of the docs: question to be
answered—redirects within the guide or redirects to old support pages?

brycekahle committed:
Add staging s3 deployment environment

jme783 committed:
Fix the build...still need to ensure the crawler actually is crawling the whole site

brycekahle committed:
Change staging deploy to feature/metalsmith while iterating a lot

brycekahle committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Bump up timeouts to 20s to hopefully make Travis pass

brycekahle committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Add slack integration to Travis

brycekahle committed:
Run optipng on all the pngs

brycekahle committed:
Run on all PNGs

brycekahle committed:
Fix travis.yml file

brycekahle committed:
Ignore .ai files when uploading

jme783 committed:
Fixed JS errors on homepage

brycekahle committed:
Convert bmp to jpg

brycekahle committed:
Change extension for .PNG images to .png

cmsunu28 committed:
updated with all the pull requests since june 12

cmsunu28 committed:
changed some images and integrated all PRs to master since june 12

cmsunu28 committed:
resolved some video and link issues

cmsunu28 committed:
attempt to fix broken link

cmsunu28 committed:
playing with footnotes on staging

chobberoni committed:
new troubleshooting page using brackets

-Updated Metalsmith page 1 loader
-Updated troubleshooting page
-General shipping info page moved to end of list

jme783 committed:
Removed monitor, fix build

cmsunu28 committed:
let's go back to :hover

brycekahle committed:
Fix link checker

brycekahle committed:
Move favicon to header

brycekahle committed:
Add missing closing div

brycekahle committed:
Fix edit links

brycekahle committed:
Add link content to crawler output

brycekahle committed:
Use current git branch in edit links so it doesn't break crawler

chobberoni committed:
updated troubleshooting

updated content

brycekahle committed:
Use branch from travis when testing in travis

brycekahle committed:
Break out success examples by operation too

Commit:
Updated Manufacturing Section

Look at me pushing code and whatnot.

cmsunu28 committed:
fixed links

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

chobberoni committed:
major section updates to troubleshooting

sections added:
Device management -
Cloud solutions -
Device Mode switching -
Installing Particle cli
Device Key management -

Commit:
I am so sad and tired

Finished the manufacturing section

chobberoni committed:
troubleshooting the troubleshooting docs; i'm sad at particle house with Will Hart :'(

left off at section -main LED off, small blue LED dim

on

http://support.particle.io/hc/en-us/sections/200785464-Troubleshoot-by-c
olor-on-the-Core

brycekahle committed:
Fix api gen TOC links

cmsunu28 committed:
fixed more links

cmsunu28 committed:
more link fixes, I think I lost some of them

brycekahle committed:
Restore datasheets and images

brycekahle committed:
Fix links

brycekahle committed:
Fix more links. Use real user-agent.

brycekahle committed:
Fix DFU links. Allow 429 status code.

Commit:
I am less tired and sad

cmsunu28 committed:
removed /assets/images/ide-firmware-versions.png from build.md

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

Commit:
More content edits.  Added a gif.

http://media.giphy.com/media/bI5BEfwbdVPcA/giphy.gif

jme783 committed:
Refactored popup css

cmsunu28 committed:
added a gif for william

Commit:
Hi Jeff!

Commit:
Merge remote-tracking branch 'origin/feature/metalsmith' into feature/metalsmith

Conflicts:
	src/content/guide/how-to-build-a-product/pcb-design.md

cmsunu28 committed:
I'm in your docs, fixin your popups

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Fix rendering of the docs on mobile screens

chobberoni committed:
my "shippable commit" before review

This is what I believe to be at the 90%-98% shippable state.

Zen desk Help button embedded in all support pages.

-Brcye, Jeff let me know if my links are broken on anything. I want to
be committed to good code pushes.

-Steph still needs to look over this?

chobberoni committed:
fixed my broken links

forgot to take out localhost pre-cursor for links
my apologies.

chobberoni committed:
this should pass Travis-CI , fingers crossed

jme783 committed:
Merged with Christine's changes

jme783 committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith

chobberoni committed:
found and tested "broken links"

chobberoni committed:
setup forward slash fix

jme783 committed:
Fixed broken link, bumped up timeout, added README support

chobberoni committed:
new content for photon factory reset

jme783 committed:
A couple small CSS tweaks

brycekahle committed:
Exclude generated api docs for now until they are ready

cmsunu28 committed:
firmware page updates

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Added a hamburger menu for showing and hiding the navigation

cmsunu28 committed:
more firmware page fixes

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

jme783 committed:
Collapse/expand the menu on page resize

jme783 committed:
don't hide the search bar until phone-size screens

jme783 committed:
hide the search when the user clicks away from it

jme783 committed:
Added search to the 404 page and made sure it worked

jme783 committed:
Add -webkit prefixes to get flexbox working with Safari

technobly committed:
Photon has SETUP and RESET buttons, rather than MODE and RST.

brycekahle committed:
API gen docs mobile css fixes

idokleinman committed:
update iOS reference for events

idokleinman committed:
iOS reference fixes

idokleinman committed:
fix h2/h3/h4 headings so TOC will look nice

m-mcgowan committed:
clarification on safe mode

Commit:
Updated a bunch of Kickstarter details

*Changed pre-order date to November
*Re-wrote backer-kit details
*Updated credit-card information

Commit:
Changed maker kit wording

Commit:
got rid of Particle is not a development platform b/c it felt icky..

jme783 committed:
Update dashboard docs to include local compiling instructions

vivek0723 committed:
Update general-shipping-info.md

cmsunu28 committed:
added intro and got rid of the floating 1 @william jeez

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

vivek0723 committed:
Update general-shipping-info.md

vivek0723 committed:
Update general-shipping-info.md

zsup committed:
Tweaks to the PCB design page

zsup committed:
Switched up the 'how to build a product' order

brycekahle committed:
Fix crawler test urls by adding a slash to the end.

zsup committed:
Content edits on the 'how to build a product' dashboard page

zsup committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith

zsup committed:
Finished edits to the 'dashboard' section

zsup committed:
Fix some links so the tests pass

zsup committed:
Edits on the manufacturing page

zsup committed:
Edits to the mobile app page

zsup committed:
Final edits to manufacturing

jme783 committed:
Added tracking for who viewed How to Build a Product with Particle. Sales ftw

jme783 committed:
Fixed bug with expanding/collapsing menu sections on page load

cmsunu28 committed:
popups and footnotes now more specific with their toggle

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

brycekahle committed:
Update dfu-util links

brycekahle committed:
SourceForge is back up, use that for dfu-util links

technobly committed:
Updated P1 Datasheet v004

technobly committed:
Updated Photon Datasheet v011

technobly committed:
Remove Trailing Whitespace in Datasheets

towynlin committed:
Newly document OAuth client creation

Also mention as coming soon:

- organization-specific clients
- customer scoped access tokens

towynlin committed:
Web app guide is shippable

Many, many "Coming Soon!" features.

brycekahle committed:
Move to explicit ordering and headers. Fix operation hrefs.

brycekahle committed:
Enable api gen again

brycekahle committed:
Don't ignore api gen docs

brycekahle committed:
Disable autolinking

Commit:
steph's recommended changes

-changed some language issues
-clarified some points

Commit:
fixed - broken datasheets link

Not sure how this was ever changed as my last commit before this
perfect and I didn’t touch this section at all.

jme783 committed:
Bumped up timeout of server test to 30 seconds

jme783 committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith

technobly committed:
Pads are on Photon, not P1

technobly committed:
table glitch

brycekahle committed:
Add docs-beta deploy target. Cache node_modules.

monkbroc committed:
Add community supported clients page

technobly committed:
Removed a couple straggler section numbers

monkbroc committed:
Document the SOS pattern

brycekahle committed:
Consolidate to one API page

jme783 committed:
Merge pull request #100 from monkbroc/sos-mode

Document the SOS pattern

jme783 committed:
Merge pull request #99 from monkbroc/community-supported-clients

Add community supported clients page

jme783 committed:
Merge pull request #97 from spark/feature/metalsmith-safe-mode

clarification on safe mode

idokleinman committed:
design H4 to look different than H3

idokleinman committed:
OMG OMG! ANDROID SDK DOCS!!

brycekahle committed:
Revert "SourceForge is back up, use that for dfu-util links"

This reverts commit 6d546d1058784846e4973b65efe61fcb5a8ef57c.

towynlin committed:
Fix WiFi.listening() docs to fit reality

vivek0723 committed:
Fixed typo in international shipping

vivek0723 committed:
Update general-shipping-info.md

jme783 committed:
Updated compiling firmware for product creators

jme783 committed:
Fixed scrollspy on API page

jme783 committed:
Added a more targeted viewport tag to ensure that mobile renders correctly

Commit:
describe noDefaults param for webhooks

jme783 committed:
Add permalink icons to all the headers! Yay!

jme783 committed:
bump up the timeout on scrolling to a section on page load to 1s

jme783 committed:
Fixed merge conflicts with master

jme783 committed:
Small fix for device dropdown menu

jme783 committed:
Added new picture for getting started with the photon

cmsunu28 committed:
added modes

cmsunu28 committed:
Merge branch 'feature/metalsmith' of https://github.com/spark/docs into feature/metalsmith

cmsunu28 committed:
removed redundant info about safe mode

kennethlimcp committed:
applicaiton --> application

jme783 committed:
Added note about selecting a device in Particle Dev

brycekahle committed:
Fix webhooks TOC

jme783 committed:
Added permalinks to api-generated docs

brycekahle committed:
Update api branch to production

jme783 committed:
Fixed merge conflict on guide build page

jme783 committed:
Merge branch 'feature/metalsmith' of github.com:spark/docs into feature/metalsmith
